influent-java/src/main/java/influent/forward/MsgPackPingDecoder.java

```diff
@@ -0,0 +1,122 @@
+package influent.forward;
+
+import org.msgpack.value.ImmutableArrayValue;
+import org.msgpack.value.ImmutableValue;
+import org.msgpack.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Optional;
+
+final class MsgPackPingDecoder {
+  private static final Logger logger = LoggerFactory.getLogger(MsgPackPingDecoder.class);
+
+  private final ForwardSecurity security;
+  private final Optional<ForwardClientNode> node;
+  private final byte[] nonce;
+  private final byte[] userAuth;
+
+  public MsgPackPingDecoder(ForwardSecurity security,
+                            Optional<ForwardClientNode> node,
+                            byte[] nonce,
+                            byte[] userAuth) {
+    this.security = security;
+    this.node = node;
+    this.nonce = nonce;
+    this.userAuth = userAuth;
+  }
+
+  public Optional<CheckPingResult> decode(ImmutableValue value) {
+    logger.debug("decoding ping {}", value);
+    String ping;
+    String selfHostname;
+    String sharedKeySalt;
+    String sharedKey = null;
+    String sharedKeyHexDigest;
+    String username;
+    String passwordDigest;
+    if (!value.isArrayValue()) {
+      error("A ping message must be an array", value);
+    }
+    ImmutableArrayValue arrayValue = value.asArrayValue();
+    if (arrayValue.size() != 6) {
+      error("A ping message must have 6 elements", value);
+    }
+    ping = decodeString(arrayValue.get(0));
+    if (!ping.equals(("PING"))) {
+      error("Invalid ping message", value);
+    }
+    selfHostname = decodeString(arrayValue.get(1));
```

Comment:
I like `clientHostname` since `self` is a little ambiguous(Influent or access client).